components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/security/authenticator/MutualSSLAuthenticator.java

```diff
@@ -0,0 +1,208 @@
+        authContext.setUsername(subjectDN);
+        try {
+            LdapName ldapDN = new LdapName(subjectDN);
+            for (Rdn rdn : ldapDN.getRdns()) {
+                if (APIConstants.CERTIFICATE_COMMON_NAME.equalsIgnoreCase(rdn.getType())) {
+                    authContext.setUsername((String) rdn.getValue());
+                }
+            }
+        } catch (InvalidNameException e) {
+            log.warn("Cannot get the CN name from certificate:" + e.getMessage());
+            authContext.setUsername(subjectDN);
+        }
+        authContext.setApiTier(apiLevelPolicy);
+        APIIdentifier apiIdentifier = getAPIIdentifier(messageContext);
+        authContext.setKeyType(APIConstants.API_KEY_TYPE_PRODUCTION);
+        authContext.setStopOnQuotaReach(true);
+        authContext.setApiKey(uniqueIdentifier + "_" + apiIdentifier.toString());
+        authContext.setTier(tier);
+        /* For the mutual SSL based authenticated request, the resource level throttling is not considered, hence
+        assigning the unlimited tier for that. */
+        VerbInfoDTO verbInfoDTO = new VerbInfoDTO();
+        verbInfoDTO.setThrottling(APIConstants.UNLIMITED_TIER);
+        messageContext.setProperty(APIConstants.VERB_INFO_DTO, verbInfoDTO);
+        if (log.isDebugEnabled()) {
+            log.debug("Auth context for the API " + getAPIIdentifier(messageContext) + ": Username[" + authContext
+                    .getUsername() + "APIKey[(" + authContext.getApiKey() + "] Tier[" + authContext.getTier() + "]");
+        }
+        APISecurityUtils.setAuthenticationContext(messageContext, authContext, null);
+    }
+
+    /**
+     * To get the API Identifier of the current API.
+     *
+     * @param messageContext Current message context
+     * @return API Identifier of currently accessed API.
+     */
+    private APIIdentifier getAPIIdentifier(MessageContext messageContext) {
+        String apiWithversion = (String) messageContext.getProperty(RESTConstants.SYNAPSE_REST_API);
+        String apiPublisher = (String) messageContext.getProperty(APIMgtGatewayConstants.API_PUBLISHER);
+        String api = null;
+        //if publisher is null,extract the publisher from the api_version
+        if (apiPublisher == null && apiWithversion != null) {
+            int ind = apiWithversion.indexOf("--");
+            apiPublisher = apiWithversion.substring(0, ind);
+        }
+
+        if (apiWithversion != null) {
+            int index = apiWithversion.indexOf("--");
+            if (index != -1) {
+                apiWithversion = apiWithversion.substring(index + 2);
+            }
+            String[] splitParts = apiWithversion.split(":");
+            api = splitParts[0];
+            apiWithversion = splitParts[1].substring(1);
+        }
+        return new APIIdentifier(apiPublisher, api, apiWithversion);
+    }
+
+    @Override
+    public String getChallengeString() {
+        return "Mutual SSL realm=\"WSO2 API Manager\"";
```

Comment:
What is this used for? Having "WSO2" hard-coded can be a security issue in case this is exposed to clients.

Reply:
When there is an API invocation failure due to authorization failure, we add a http header "WWW-Authenticate" with this value from authenticator. For our OAuthAuthenticator, we have "OAuth2 realm=\"WSO2 API Manager\"", I followed the same pattern.  If we are to change this, shall we add "Mutual SSL" only?

Reply:
Shall we read the product name from the carbon.xml and use it here?

Reply:
Sure.